src/VisualStudio/Core/Def/ProjectSystem/SdkAnalyzerAssemblyRedirector.cs

```diff
@@ -0,0 +1,210 @@
+        {
+            return Path.GetFullPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, @"CommonExtensions\Microsoft\DotNet"));
+        }
+
+        return null;
+    }
+}
+
+/// <summary>
+/// Core functionality of <see cref="SdkAnalyzerAssemblyRedirector"/> extracted for testing.
+/// </summary>
+internal class SdkAnalyzerAssemblyRedirectorCore : IAnalyzerAssemblyRedirector
+{
+    private readonly IVsActivityLog? _log;
+
+    /// <summary>
+    /// Map from analyzer assembly name (file name without extension) to a list of matching analyzers.
+    /// </summary>
+    private readonly Lazy<ImmutableDictionary<string, List<AnalyzerInfo>>> _analyzerMap;
+
+    public SdkAnalyzerAssemblyRedirectorCore(string? insertedAnalyzersDirectory, IVsActivityLog? log = null)
+    {
+        _log = log;
+        _analyzerMap = new(() => CreateAnalyzerMap(insertedAnalyzersDirectory));
+    }
+
+    private ImmutableDictionary<string, List<AnalyzerInfo>> CreateAnalyzerMap(string? insertedAnalyzersDirectory)
+    {
+        if (insertedAnalyzersDirectory == null)
+        {
+            Log("Analyzer redirecting is disabled.");
+            return ImmutableDictionary<string, List<AnalyzerInfo>>.Empty;
+        }
+
+        var metadataFilePath = Path.Combine(insertedAnalyzersDirectory, "metadata.json");
+        if (!File.Exists(metadataFilePath))
+        {
+            Log($"File does not exist: {metadataFilePath}");
+            return ImmutableDictionary<string, List<AnalyzerInfo>>.Empty;
+        }
+
+        Dictionary<string, string>? versions;
+
+        try
+        {
+            versions = JsonSerializer.Deserialize<Dictionary<string, string>>(File.ReadAllText(metadataFilePath));
+        }
+        catch (Exception ex)
+        {
+            Log($"Failed to read or parse metadata file '{metadataFilePath}'. {ex}", __ACTIVITYLOG_ENTRYTYPE.ALE_WARNING);
+            return ImmutableDictionary<string, List<AnalyzerInfo>>.Empty;
+        }
+
+        if (versions is null || versions.Count == 0)
+        {
+            Log($"Versions are empty: {metadataFilePath}", __ACTIVITYLOG_ENTRYTYPE.ALE_WARNING);
+            return ImmutableDictionary<string, List<AnalyzerInfo>>.Empty;
+        }
+
+        var builder = ImmutableDictionary.CreateBuilder<string, List<AnalyzerInfo>>(StringComparer.OrdinalIgnoreCase);
+
+        // Expects layout like:
+        // VsInstallDir\DotNetRuntimeAnalyzers\WindowsDesktopAnalyzers\analyzers\dotnet\System.Windows.Forms.Analyzers.dll
+        //                                     ~~~~~~~~~~~~~~~~~~~~~~~                                                     = topLevelDirectory
+        //                                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ = analyzerPath
+
+        foreach (var topLevelDirectory in Directory.EnumerateDirectories(insertedAnalyzersDirectory))
```

Comment:
Is there a way we can produce a file that produces all this up front so we don't need to search for binaries that ship in the product? This is having the user pay a cost for things we can absolutely determine because we insert these binaries.